jobrunner/run.py

```diff
@@ -195,28 +161,161 @@ def handle_running_job(job):
             cleanup_job(job)
 
 
-def handle_running_job_api(job, api):
+def handle_active_jobs_api(api, active_jobs):
+    for job in active_jobs:
+        # `set_log_context` ensures that all log messages triggered anywhere
+        # further down the stack will have `job` set on them
+        with set_log_context(job=job):
+            try:
+                handle_job_api(job, api)
+            except Exception:
+                mark_job_as_failed(job, f"Internal error")
+                # Do not clean up, as we may want to debug
+                #
+                # Raising will kill the main loop, by design. The service manager
+                # will restart, and this job will be ignored when it does, as
+                # it has failed. If we have an internal error, a full restart
+                # might recover better.
+                raise
+
+
+# we do not control the tranisition from these states, the executor does
+STABLE_STATES = [
+    ExecutorState.PREPARING,
+    ExecutorState.EXECUTING,
+    ExecutorState.FINALIZING,
+]
+
+
+def handle_job_api(job, api):
+    """Handle an active job.
+
+    This contains the main state machine logic for a job. For the most part,
+    state transitions follow the same logic, which is abstracted. Some
+    transitions require special logic, mainly the initial and final states, as
+    well as supporting cancellation.
+    """
+    assert job.state in (State.PENDING, State.RUNNING)
+    definition = job_to_job_definition(job)
+
     if job.cancelled:
-        log.info("Cancellation requested, killing job")
-        api.terminate(job_to_job_definition(job))
+        # cancelled is driven by user request, so is handled explicitly first
+        # regardless of executor state.
+        api.terminate(definition)
+        api.cleanup(definition)
+        mark_job_as_failed(job, "Cancelled by user", StatusCode.CANCELLED_BY_USER)
+        return
 
-    try:
-        is_running = sync_job_status(job, api)
+    old_status = api.get_status(definition)
```

Comment:
I keep stumbling over this name `old_status`, even though I know it's been like this for ages. Would you consider "initial" or "before" as an alternative to "old"? The latter suggests to me that it's somehow out-of-date (or may be), but actually it's the very latest status.

Reply:
`initial` works for me. I originally used `current`, which did NOT work.